### PR TITLE
CSE-434 Implement tests for SPIFlashFileSystem library

### PR DESCRIPTION
**Pull request by umlsynco**
1. Add api tests and functional test                  
2. Fix BAD sector detection issue                    
3. Fix issue with File.read API. Return blob-payload with seek position 0.

### CODE REVIEW

File: SPIFlashFileSystem.class.nut [hunk 1]
```diff
@@ -1027,7 +1029,8 @@ class SPIFlashFileSystem.File {
     function read(len = null) {
         local data = _filesystem._read(_fileId, _pos, len);
         _pos += data.len();
-
+        // make data blob ready to read
+        data.seek(0);
```
Comment (snork-nbl) [inline SPIFlashFileSystem.class.nut:1048]: Good catch

File: tests/BaseSFFS.device.nut [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  sffs = null;
+  pages = 0;
+  start = 0;
+
+  //
+  // Initialization of the SpiFlashFileSystem library
+  //  - create an SpiFlashFileSystem class object
+  //  - erase all SPI flash data
+  //  - initialize a new FAT block
+  function setUp() {
```
Comment (snork-nbl) [inline tests/BaseSFFS.device.nut]: indentations should be 4 spaces https://electricimp.com/docs/squirrel/codestyle/#indentations

File: tests/BaseSFFS.device.nut [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    return Promise(function(ok, err) {
+      // check that we're on  003+
+
+      this.assertTrue("spiflash" in hardware, "imp003 and above is expected");
+
+      // get actual flash size
+      hardware.spiflash.enable();
+      this.size = hardware.spiflash.size();
```
Comment (snork-nbl) [inline tests/BaseSFFS.device.nut]: it doesn't hurt, but there is no need to use `this` here

File: SPIFlashFileSystem.class.nut [hunk 4]
```diff
@@ -123,10 +123,11 @@ class SPIFlashFileSystem {
         if (!_fat.fileExists(fname)) throw ERR_FILE_NOT_FOUND;
         if (isFileOpen(fname)) throw ERR_OPEN_FILE;
 
-        _enable();
-
         // Build a blob to zero out the file
         local zeros = blob(SPIFLASHFILESYSTEM_HEADER_SIZE + SPIFLASHFILESYSTEM_TIMESTAMP_SIZE + SPIFLASHFILESYSTEM_MAX_FNAME_SIZE + 1);
+
```
Comment (snork-nbl) [inline SPIFlashFileSystem.class.nut:130]: Why is this change needed?

File: tests/SffsFatInternalApi.device.test.nut [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    local addr = sffs._fat.getFreePage();
+    return Promise(function(ok, err) {
+      imp.wakeup(2, function() {
+        try {
+          local flags = [SPIFLASHFILESYSTEM_STATUS_FREE,
+            SPIFLASHFILESYSTEM_STATUS_ERASED,
+            SPIFLASHFILESYSTEM_STATUS_USED,
+            SPIFLASHFILESYSTEM_STATUS_BAD
+          ];
+          for (local i = 0; i < 4; ++i) {
```
Comment (snork-nbl) [inline tests/SffsFatInternalApi.device.test.nut]: `flags.length` would probably be better here

File: tests/SffsFileApi.device.test.nut [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    buffer = "0123456789012345678901234567890123456789";
+    // max iterations to ocupay all pages by a single file
+    iterations = this.pages * SPIFLASHFILESYSTEM_SECTOR_SIZE / buffer.len() - 10;
+    maxHeaderSize = SPIFLASHFILESYSTEM_HEADER_SIZE +
+      SPIFLASHFILESYSTEM_TIMESTAMP_SIZE;
+  }
+
+    //
+    // Check that max size file
+    function test01_writeMaxPagesFile() {
```
Comment (snork-nbl) [inline tests/SffsFileApi.device.test.nut]: There is something wrong with indentation here

File: tests/SffsFileApi.device.test.nut [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        this.assertTrue(file2.len() < expectedSize);
+      }
+    }
+    file1.close();
+    file2.close();
+  }
+
+  // Check filesize after the previous test which
+  // should be finished with exception
+  function test07_checkFileSize() {
```
Comment (snork-nbl) [inline tests/SffsFileApi.device.test.nut]: Usually it's not the best idea to make unit tests dependent on the order of execution. It would probably make sense to make this test a part of the previous one.

File: tests/BaseSFFS.device.nut [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  //
+  // Initialization of the SpiFlashFileSystem library
+  //  - create an SpiFlashFileSystem class object
+  //  - erase all SPI flash data
+  //  - initialize a new FAT block
+  function setUp() {
+    return Promise(function(ok, err) {
+      // check that we're on  003+
+
+      this.assertTrue("spiflash" in hardware, "imp003 and above is expected");
```
Comment (sergey-nbl) [inline tests/BaseSFFS.device.nut]: How does this affect Promise state?

File: tests/BaseSFFS.device.nut [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+      // get actual flash size
+      hardware.spiflash.enable();
+      this.size = hardware.spiflash.size();
+      hardware.spiflash.disable();
+      // number of the available pages
+      this.pages = this.size / 4096;
```
Comment (sergey-nbl) [inline tests/BaseSFFS.device.nut]: Magic number should be used at the code

File: tests/SFFS.device.test.nut [hunk 10]
```diff
@@ -1,247 +1,260 @@
...
+    this.sffs.open("file4.txt", "w").close();
+    this.assertClose(time(), this.sffs.created("file4.txt"), 1);
+  }
+
+  //
+  // Helper method to test filename input parameters
+  //
+  function _openTest(filename) {
+    try {
+      sffs.open(null, "w");
```
Comment (sergey-nbl) [inline tests/SFFS.device.test.nut]: Here should be filename, I guess

File: tests/SffsFileApi.device.test.nut [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    this.assertTrue(file.eof());
+  }
+
+
+  // seek, tell eof after file close
+  function test05_fileClose() {
+    local file = sffs.open("test.txt", "r");
+    file.close();
+
+    this.assertEqual(file.tell(), 0);
```
Comment (sergey-nbl) [inline tests/SffsFileApi.device.test.nut]: Duplicates previous test

File: tests/SffsFileApi.device.test.nut [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+      // it should be possible to seek at the same position
+      file.seek(0);
+      seekPass = true;
+      // check that possition is correct
+      assertEqual(0, file.tell());
+      // this seek should throw an exception
+      file.seek(1);
+    } catch (e) {
+      assertEqual(seekPass, true,
+        "Expected that it is possible to seek on 0 for an empt file");
```
Comment (sergey-nbl) [inline tests/SffsFileApi.device.test.nut]: typo "empt"

File: SPIFlashFileSystem.class.nut (hunk 4 above)
Comment (umlsynco) [reply SPIFlashFileSystem.class.nut:130]: There is no special reason for that except that _enable() method allocate "potentially critical" resource AND blob creation does not required that resource.

So, please let me know if it could be a problem with merge of this change on the main branch ?

File: tests/BaseSFFS.device.nut (hunk 9 above)
Comment (umlsynco) [reply tests/BaseSFFS.device.nut]: Agree, I'll change it

File: tests/BaseSFFS.device.nut (hunk 3 above)
Comment (umlsynco) [reply tests/BaseSFFS.device.nut]: Can I change "this.size" on "size" on the callback methods, when I make bindenv(this) for a callback method ?

File: tests/SFFS.device.test.nut (hunk 10 above)
Comment (umlsynco) [reply tests/SFFS.device.test.nut]: As you say "good catch" :)  Thanks a lot

File: tests/SffsFatInternalApi.device.test.nut (hunk 5 above)
Comment (umlsynco) [reply tests/SffsFatInternalApi.device.test.nut]: changed

File: tests/SffsFileApi.device.test.nut (hunk 11 above)
Comment (umlsynco) [reply tests/SffsFileApi.device.test.nut]: The previous one was for a new file, but this one if for reading of an existing file.

And one more interesting aspect of this test is that we have close previous test with file seek position equal to the eof.  And we double check that a new instance of the opened file has a correct seek position.

File: tests/SffsFileApi.device.test.nut (hunk 7 above)
Comment (umlsynco) [reply tests/SffsFileApi.device.test.nut]: Good comment. 
Agree, these tests were separated because I've used Promise originally.

I'll change it

File: SPIFlashFileSystem.class.nut (hunk 4 above)
Comment (snork-nbl) [reply SPIFlashFileSystem.class.nut:130]: Got it. Thanks. I don't see a problem here.

File: tests/BaseSFFS.device.nut (hunk 3 above)
Comment (snork-nbl) [reply tests/BaseSFFS.device.nut]: Yes you can.

**Review by snork-nbl — APPROVED** [review]
